### ISSUE TITLE
`Metric`s should be related to the task/dataset and not to the `Model`, maybe?

### ISSUE THREAD

**Issue by bryant1410** [allenai/allennlp#3824]
**Is your feature request related to a problem? Please describe.**

I find myself creating instances of the same `Metric`s when I build `Model`s for the same `DatasetReader`, so I wonder if they should be closer to the "task" or "dataset" than to the model.

I wanted to hear opinions on this at least.

**Describe the solution you'd like**

I know this is a breaking change, but somehow `Metric`, IMHO should be closer to the data. Maybe to `DatasetReader` (and `Model.forward` can be expected to return `scores` as well as the `loss`)? I'm not convinced, because it should just read the data. However, it knows about the kind of data. Like in BiDAF, it says `squad_metrics`: https://github.com/allenai/allennlp-reading-comprehension/blob/e4f8e5df4f9fa35287d44e94fc8b26b9cabed0a5/allennlp_rc/models/bidaf.py#L329-L337

**Describe alternatives you've considered**
An alternative could be to have a "base" model that defines the metrics, but still, I think we metrics are related to the task/problem. Like in SQuAD, you care about F1 and ExactMatch, regardless of the model.

**Comment by bryant1410** [comment]
Maybe the `Trainer` (which already does a lot of stuff..)?

**Comment by matt-gardner** [comment]
Let's say the `Metric` lives with the `DatasetReader`.  How does it actually get used?

**Comment by bryant1410** [comment]
An iterable or dict of `Metric`s are obtained from the dataset from `TrainerModel` and passed to the `Trainer`, which does gradient descent, calling `Model.forward` for each batch, getting the `predictions` (a tensor; or the raw `scores`) as well as the `loss`, passing the former to each `Metric`, and then computing the value from each `Metric` (which preferably IMHO should return a `Dict[str, float]` so it's a homogeneous API).

**Comment by matt-gardner** [comment]
I guess I don't really see what's gained by making this change.  One thing that's lost is that the `Model` now has more constraints on its output, instead of just having to return a `loss`.  How do you know which key to pass predictions in?  And what if you want different metrics that depend on different predictions?  How do you handle the mapping?  What if you want metrics on intermediate model computations?  Seems strictly less flexible to do what you propose, and we do have concrete uses for all of what I said above, which would be hard in this setting.

**Comment by bryant1410** [comment]
The dataset can have a `get_metrics() -> Iterable[Metric]` method but also another method like `get_keys() -> Iterable[str]` (or a better name) which could be, for example, `["scores", "probabilities", "answer_string", "start_span_index"]`. Then, any model that consumes a dataset has to return those (+ "loss") in the `forward` method. The model already has a strong dependency on the dataset anyway, because of what expects as parameters in the mentioned method. And the trainer can pass those outputs to the dataset, which knows about the metrics. Maybe the `get_keys` is not needed, can be just stated in the docstrings, but a model that uses that dataset knows about those, so the trainer gets them from `forward` and passes them to the dataset to compute the metrics and get the results from it. I'm writing ideas... Maybe then it could just be like `Dataset.get_metric_values(model_output: Dict[str, Any], reset: bool) -> Dict[str, float]`

Benefits:

* Easier to use generic models (e.g., `BasicClassifier`, `SimpleTagger`, the former `BertForClassification`, whatever custom one a user makes) as they can adapt to any metric, given that they return the expected key (e.g, `predictions` or `scores`; which can be standard for vanilla classification). An alternative here is adding a param `metrics` to them, so you define them in the params config file. Another, worse IMO, is to subclass these models.
* Not need to define the metrics again for different models of the same dataset.
* It makes more sense IMHO, design-wise, to have the metrics closer to the dataset than to the model.

Does it make sense?

**Comment by bryant1410** [comment]
> Maybe then it could just be like `Dataset.get_metric_values(model_output: Dict[str, Any], reset: bool) -> Dict[str, float]`

The dataset here is kinda a leaderboard server: it provides the data to you, you give your model output and it tells you your score :smile:

**Comment by matt-gardner** [comment]
So, I don't think this is something we want to do.  We want to keep our abstractions separated, into data, model, and training.  This adds more of the coupling that we intentionally avoided when designing this library.  It also doesn't really have a satisfying answer to the issues that I brought up - if you're saying that the data has to specify keys for which you might want to compute model-internal metrics, then you're just saying that I need one dataset reader per model.  I also don't really see what's not generic about how metrics work with our current classifiers and taggers.

**Comment by bryant1410** [comment]
The models are already coupled to the dataset, right? For example, `BasicClassifier` expects a dataset reader that provides something called "tokens" and something called "labels" (so the dataset reader has to return these names already), and nothing else. If you have some dataset that returns slightly different names, or that returns more fields, the execution will crash. And this model already returns "logits" and "probs". I think there won't be more coupling that what already exists.

What's not generic about the metrics is that, for example, I may want to measure F1 or weighted F1, or whatever other measures, instead of accuracy in `BasicClassifier`, and the way to do that AFAIK is to either subclass it or to copy-paste some of it.

**Comment by bryant1410** [comment]
I think this is a common thing when people release PyTorch datasets: they provide a dataset on one side and then metrics on the other side. So in the end, they're coupled to some extent.

**Comment by matt-gardner** [comment]
The model is *needs to match* the dataset reader in our default training loop, but it doesn't ever actually get *access* to the dataset reader object.  The latter would be coupling the two objects.  Similarly with the trainer - it doesn't ever actually have access to any dataset reader, only the result of calling one.  This means that you can use our models and trainers with other ways of producing data, if you want to.  If we do what you suggest, the actual dataset reader object would have to be shared and passed around in really messy ways.

You're right that conceptually there are a set of metrics that are tied to particular datasets. But from a software design perspective, it makes a lot more sense to attach a metric to a model, especially as the model is already typically specific to a dataset.

A solution to the generic metric problem that you mention is to make them configurable in a `Model's` constructor.  We debated about doing this, having `Metric` be `Registrable` and configurable in that way.  The problem is that there's no unified `Metric` API, because different metrics need different inputs, so it doesn't make sense to have a completely generic `Registrable` `Metric` (hmm, actually, it looks like we *did* implement this, though we don't actually use it anywhere, for the reason above).  But you could still imagine having a classification model have a list of allowable metrics, and a configuration option that allows a user to select them.  This would be totally fine.
